trio/_core/_run.py

```diff
@@ -1446,11 +1455,126 @@ def task_exited(self, task, outcome):
             if self.tasks:  # pragma: no cover
                 raise TrioInternalError
         else:
+            if task is self.main_task:
+                self.main_task_outcome = outcome
+                outcome = Value(None)
             task._parent_nursery._child_finished(task, outcome)
 
         if self.instruments:
             self.instrument("task_exited", task)
 
+    ################
+    # Async generator finalization support
+    ################
+
+    async def finalize_asyncgen(self, agen, name, *, check_running):
+        if check_running and agen.ag_running:
+            # Another async generator is iterating this one, which is
+            # suspended at an event loop trap. Add it back to the
+            # asyncgens set and we'll get it on the next round.  Note
+            # that this is only possible during end-of-run
+            # finalization; in GC-directed finalization, no one has a
+            # reference to agen anymore, so no one can be iterating it.
+            #
+            # This field is only reliable on 3.8+ due to
+            # ttps://bugs.python.org/issue32526. Pythons below
+            # 3.8 use a workaround in finalize_remaining_asyncgens.
```

Comment:
Missing `h` in the link.

Also, I'm not sure why we need the `check_running` flag – on new Python's, the `check_running` is always true, and on old Pythons, `ag_running` will always be False, right? So `if agen.ag_running: ...` would be equivalent.

Reply:
Good point. It was originally trying to avoid a more expensive analysis for older Pythons in the GC-driven case (since the asyncgen can't possibly be running when it's GCed), but I removed that when I noticed the simpler workaround.